examples/ecs/computed_states.rs

```diff
@@ -0,0 +1,682 @@
+//! This example illustrates the use of [`ComputedStates`] for more complex state handling patterns.
+//!
+//! In this case, we'll be implementing the following pattern:
+//! - The game will start in a `Menu` state, which we can return to with `Esc`
+//! - From there, we can enter the game - where our bevy symbol moves around and changes color
+//! - While in game, we can pause and unpause the game using `Space`
+//! - We can also toggle "Turbo Mode" with the `T` key - where the movement and color changes are all faster. This
+//!   is retained between pauses, but not if we exit to the main menu.
+//!
+//! In addition, we want to enable a "tutorial" mode, which will involve it's own state that is toggled in the main menu.
+//! This will display instructions about movement and turbo mode when in game and unpaused, and instructions on how to unpause when paused.
+//!
+//! To implement this, we will create 2 root-level states: [`AppState`] and [`TutorialState`].
+//! We will then create some computed states that derive from [`AppState`]: [`InGame`] and [`TurboMode`] are marker states implemented
+//! as Zero-Sized Structs (ZSTs), while [`IsPaused`] is an enum with 2 distinct states.
+//! And lastly, we'll add [`Tutorial`], a computed state deriving from [`TutorialState`], [`InGame`] and [`IsPaused`], with 2 distinct
+//! states to display the 2 tutorial texts.
+
+use bevy::prelude::*;
+
+// To begin, we want to define our state objects.
+#[derive(Debug, Clone, Copy, Default, Eq, PartialEq, Hash, States)]
+enum AppState {
+    #[default]
+    Menu,
+    // Unlike in the `states` example, we're adding more data in this
+    // version of our AppState. In this case, we actually have
+    // 4 distinct "InGame" states - unpaused and no turbo, paused and no
+    // turbo, unpaused and turbo and paused and turbo.
+    InGame {
+        paused: bool,
+        turbo: bool,
+    },
+}
+
+// The tutorial state object, on the other hand, is a fairly simple enum.
+#[derive(Debug, Clone, Copy, Default, Eq, PartialEq, Hash, States)]
+enum TutorialState {
+    #[default]
+    Active,
+    Inactive,
+}
+
+// Because we have 4 distinct values of `AppState` that mean we're "InGame", we're going to define
+// a separate "InGame" type and implement `ComputedStates` for it.
+// This allows us to only need to check against one type
+// when otherwise we'd need to check against multiple.
+#[derive(Debug, Clone, Copy, Eq, PartialEq, Hash)]
+struct InGame;
```

Comment:
I think having data in the InGame struct is more idiomatic when trying to do enum variant types (which is kind of what we're doing here).

```rust
enum AppState {
    #[default]
    Menu,
    InGame(InGame),
}

struct InGame {
    paused: bool,
    turbo: bool,
}
```

Reply:
@dmyyy - In this particular case, I worry it'd add confusion to things - as well as make the computed matches much more wordy. Unless this is a strongly held opinion I'd prefer to keep this element as it is.